Take all screenshots

mokagio opened the pull request:
Adds the code to take all the screenshots on the Espresso side as well as a Fastlane lane to take them.

### To test

Run

```
bundle install install --with-screenshots \
  && bundle exec fastlane screenshots \
  locales:en,it # this is so they don't take forever, feel free to use any locale
```

and verify that the screenshots come out as expected in `fastlane/screenshots/en/images/phoneScreenshots`

![abc-screenshot-2020-06-03-212247](https://user-images.githubusercontent.com/1218433/83631160-69880280-a5e0-11ea-9aa4-139f25dbeb44.png)

---

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mokagio committed:
Add Fastlane lane to take screenshots

mokagio committed:
Take all the screenshots

mokagio committed:
Add logic to switch between light and dark theme

mokagio committed:
Remove commented code from Screen.kt

Thanks @loremattei for pointing this out.

See https://github.com/woocommerce/woocommerce-android/pull/2365#discussion_r422894125

mokagio committed:
Add status bar customization logic in screenshots tests

mokagio commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/screenshots/mystore/settings/SettingsScreen.kt:
```diff
@@ -46,4 +47,16 @@ class SettingsScreen : Screen {
+
+    fun switchToLightTheme(): MyStoreScreen {
+        clickOn(THEME_BUTTON)
+        clickOnViewWithText("Light")
```
This code selects the option for the light theme in the theme switcher.

Using the layout inspector, I couldn't find any specific ids for the options. Looking at the code, I can see its build programmatically. I need a way to reference this that is agnostic of the language.

The approach I had previously was to take all the screenshots in both theme modes, and then only keep the ones we care about from the dark and light set. That approach did work, but was quite slow as it had to have code to change the UI mode in the emulator, restart it, then wait for it to become online. It would be awesome if we could instead simply switch the theme from the app like here.

@AmandaRiu I can see you build the theme switcher. Have you got any tips? Thanks a lot!

AmandaRiu replied:
@mokagio are you saying the code above doesn't work? It seems like it should...

AmandaRiu replied:
@mokagio I originally answered on a really small screen, and now that I looked on my computer I'm seeing the actual question is regarding:

> I need a way to reference this that is agnostic of the language.

You should be able to do it the same way we set those values...using localized string resources. Here's how the options are set in the UI:

```
enum class ThemeOption(@StringRes val label: Int) {
    LIGHT(R.string.settings_app_theme_option_light),
    DARK(R.string.settings_app_theme_option_dark),
    DEFAULT(R.string.settings_app_theme_option_default)
}
```

Does that help?

mokagio replied:
YES!

![ninja-turtles-high-five](https://user-images.githubusercontent.com/1218433/83628545-04321280-a5dc-11ea-9851-7fbb1aaaa292.gif)

Using `R.string....` does the trick. TIL you can do that

mokagio committed:
Support all locales when switching theme in screenshots test

mokagio commented on fastlane/Fastfile:
```diff
@@ -394,4 +489,8 @@ end
\ No newline at end of file
+end
+
+def screenshots_directory
+  File.join(Dir.pwd, "screenshots")
+end
```
It's handy to have this extracted here because we'll reuse it later when doing the composition.

mokagio commented on fastlane/actions/take_android_emulator_screenshots.rb:
```diff
@@ -0,0 +1,133 @@
+      TAKE_ANDROID_EMULATOR_SCREENSHOTS_CUSTOM_VALUE = :TAKE_ANDROID_EMULATOR_SCREENSHOTS_CUSTOM_VALUE
+    end
+
+    class TakeAndroidEmulatorScreenshotsAction < Action
```
This and the rest of the custom Fastlane code comes from the WordPress Android codebase.

I think most of it is sort of unnecessary, but it's handy to have it because it handles booting up the emulator, which may come handy in CI.

mokagio commented on fastlane/Fastfile:
```diff
@@ -309,6 +310,100 @@ end
+      locales = options[:locales].split(',')
+    else
+      locales = SUPPORTED_LOCALES.
+        # Depending on how automated this is, possibly on whether we wipe the
+        # emulator before running or not, the zh-CN locale might break due to
+        # Pinwin Keyboard wanting contacts access. If that's the case, you can
+        # skip it by uncommeting the following line.
+        #reject { |locale| locale[:glotpress] == 'zh-cn' }.
```
Actually... I haven't seen the issue in a while. Maybe since moving to API 29?

Do you think this comment is still useful, or should I get rid of it?

jkmassel replied:
If this issue is resolved it would be super great to get rid of this comment yes!

jkmassel reviewed (changes requested):
I think this branch might be out of sync with the secret store – I kept getting a "wrong password" error until I ran `bundle exec run configure_update`.

I'm running into a couple of consistent issues:

```
[12:29:33]: ▸ com.woocommerce.android.screenshots.ScreenshotTest:
[12:29:33]: ▸ Error in screenshots(com.woocommerce.android.screenshots.ScreenshotTest):
[12:29:33]: ▸ junit.framework.AssertionFailedError: Unable to continue – expectation wasn't satisfied quickly enough
[12:29:33]: ▸ 	at junit.framework.Assert.fail(Assert.java:50)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:42)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:33)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForConditionToBeTrue(Screen.kt:271)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForElementToBeDisplayed(Screen.kt:216)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.<init>(Screen.kt:46)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.mystore.MyStoreScreen.<init>(MyStoreScreen.kt:25)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.mystore.settings.SettingsScreen.goBackToMyStoreScreen(SettingsScreen.kt:28)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.mystore.settings.SettingsScreen.switchToLightTheme(SettingsScreen.kt:54)
[12:29:33]: ▸ 	at com.woocommerce.android.screenshots.ScreenshotTest.screenshots(ScreenshotTest.kt:81)
[12:29:33]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[12:29:33]: ▸ 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[12:29:33]: ▸ 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[12:29:33]: ▸ 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[12:29:33]: ▸ 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[12:29:33]: ▸ 	at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
[12:29:33]: ▸ 	at tools.fastlane.screengrab.locale.LocaleTestRule$1.evaluate(LocaleTestRule.java:32)
[12:29:33]: ▸ 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
[12:29:33]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
[12:29:33]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:29:33]: ▸ 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:29:33]: ▸ 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
[12:29:33]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:128)
[12:29:33]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:27)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:29:33]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:29:33]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
[12:29:33]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
[12:29:33]: ▸ 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
[12:29:33]: ▸ 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:392)
[12:29:33]: ▸ 	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
[12:29:34]: ▸ Time: 175.087
[12:29:34]: ▸ There was 1 failure:
[12:29:34]: ▸ 1) screenshots(com.woocommerce.android.screenshots.ScreenshotTest)
[12:29:34]: ▸ junit.framework.AssertionFailedError: Unable to continue – expectation wasn't satisfied quickly enough
[12:29:34]: ▸ 	at junit.framework.Assert.fail(Assert.java:50)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:42)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:33)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForConditionToBeTrue(Screen.kt:271)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForElementToBeDisplayed(Screen.kt:216)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.<init>(Screen.kt:46)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.mystore.MyStoreScreen.<init>(MyStoreScreen.kt:25)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.mystore.settings.SettingsScreen.goBackToMyStoreScreen(SettingsScreen.kt:28)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.mystore.settings.SettingsScreen.switchToLightTheme(SettingsScreen.kt:54)
[12:29:34]: ▸ 	at com.woocommerce.android.screenshots.ScreenshotTest.screenshots(ScreenshotTest.kt:81)
[12:29:34]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[12:29:34]: ▸ 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[12:29:34]: ▸ 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[12:29:34]: ▸ 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[12:29:34]: ▸ 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[12:29:34]: ▸ 	at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
[12:29:34]: ▸ 	at tools.fastlane.screengrab.locale.LocaleTestRule$1.evaluate(LocaleTestRule.java:32)
[12:29:34]: ▸ 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
[12:29:34]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
[12:29:34]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:29:34]: ▸ 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:29:34]: ▸ 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
[12:29:34]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:128)
[12:29:34]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:27)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:29:34]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:29:34]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
[12:29:34]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
[12:29:34]: ▸ 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
[12:29:34]: ▸ 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:392)
[12:29:34]: ▸ 	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
[12:29:34]: ▸ FAILURES!!!
[12:29:34]: ▸ Tests run: 1,  Failures: 1
```

```
[12:39:06]: ▸ junit.framework.AssertionFailedError: Unable to continue – expectation wasn't satisfied quickly enough
[12:39:06]: ▸ 	at junit.framework.Assert.fail(Assert.java:50)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:42)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:33)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForConditionToBeTrue(Screen.kt:271)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForElementToBeDisplayed(Screen.kt:216)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.<init>(Screen.kt:46)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.orders.SingleOrderScreen.<init>(SingleOrderScreen.kt:11)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.orders.OrderListScreen.selectOrder(OrderListScreen.kt:21)
[12:39:06]: ▸ 	at com.woocommerce.android.screenshots.ScreenshotTest.screenshots(ScreenshotTest.kt:68)
[12:39:06]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[12:39:06]: ▸ 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[12:39:06]: ▸ 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[12:39:06]: ▸ 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[12:39:06]: ▸ 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[12:39:06]: ▸ 	at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
[12:39:06]: ▸ 	at tools.fastlane.screengrab.locale.LocaleTestRule$1.evaluate(LocaleTestRule.java:32)
[12:39:06]: ▸ 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
[12:39:06]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
[12:39:06]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:39:06]: ▸ 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:39:06]: ▸ 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
[12:39:06]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:128)
[12:39:06]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:27)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:39:06]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:39:06]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
[12:39:06]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
[12:39:06]: ▸ 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
[12:39:06]: ▸ 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:392)
[12:39:06]: ▸ 	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
[12:39:06]: ▸ FAILURES!!!
[12:39:06]: ▸ Tests run: 1,  Failures: 1
```

```
[12:55:08]: ▸ junit.framework.AssertionFailedError: Unable to continue – expectation wasn't satisfied quickly enough
[12:55:08]: ▸ 	at junit.framework.Assert.fail(Assert.java:50)
[12:55:08]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:42)
[12:55:08]: ▸ 	at com.woocommerce.android.screenshots.util.TestIdler.idleUntilReady(TestIdler.kt:33)
[12:55:08]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForConditionToBeTrue(Screen.kt:271)
[12:55:08]: ▸ 	at com.woocommerce.android.screenshots.util.Screen.waitForElementToBeDisplayed(Screen.kt:216)
[12:55:08]: ▸ 	at com.woocommerce.android.screenshots.mystore.StatsComponent.waitForGraphToLoad(StatsComponent.kt:27)
[12:55:08]: ▸ 	at com.woocommerce.android.screenshots.mystore.StatsComponent.switchToStatsDashboardYearsTab(StatsComponent.kt:38)
[12:55:08]: ▸ 	at com.woocommerce.android.screenshots.ScreenshotTest.screenshots(ScreenshotTest.kt:63)
[12:55:08]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[12:55:08]: ▸ 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[12:55:08]: ▸ 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[12:55:08]: ▸ 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[12:55:08]: ▸ 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[12:55:08]: ▸ 	at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
[12:55:08]: ▸ 	at tools.fastlane.screengrab.locale.LocaleTestRule$1.evaluate(LocaleTestRule.java:32)
[12:55:08]: ▸ 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
[12:55:08]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
[12:55:08]: ▸ 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:55:08]: ▸ 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:55:08]: ▸ 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
[12:55:08]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:128)
[12:55:08]: ▸ 	at org.junit.runners.Suite.runChild(Suite.java:27)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[12:55:08]: ▸ 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[12:55:08]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
[12:55:08]: ▸ 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
[12:55:08]: ▸ 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
[12:55:08]: ▸ 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:392)
[12:55:08]: ▸ 	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
```
One last note – for screenshot PR reviews I _highly_ recommend running all of the languages – otherwise it's possible that one language is just broken (`ru` often does this...) and we might miss it! :)

mokagio commented:
Hey @jkmassel, just wanted to say thanks and that I haven't forgotten about this PR. Between hack week last week and a total focus on Danger this week, I won't get time to work on you fixes till next week. Thanks for your patience.

mokagio committed:
Remove comment on filtering locales from Fastfile

Since the issue seems to be resolved (thanks to API 29 maybe?) and
filtering an array of hashes should be an easy task for anyone with
basic Ruby skills, the comment was redundant.

Thanks @jkmassel.

See https://github.com/woocommerce/woocommerce-android/pull/2498#discussion_r436059348

mokagio commented:
This now works on my end. @jkmassel can I ask you to have another look?

Although, the `tr` locale fails when running from Fastlane (but not from Android studio with `-testLocale 'tr'`). It looks like a timeout fetching the graph data.

I'm not sure whether that's a blocker or not, because in order to make the screenshots look good, we need to use fake data anyways, which removes the failure. Because of that, I haven't looked into the network failure yet.

I haven't decided how to track the fake data yet, maybe a filed guide? Open to suggestions 😄 

Anyways, here's the code I've been using:

<details>
<summary>Show diff</summary>

```diff
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt
index b3d4582c8..0a596ca24 100644
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt
@@ -26,6 +26,9 @@ import kotlinx.android.synthetic.main.fragment_dashboard.*
 import kotlinx.android.synthetic.main.fragment_dashboard.view.*
 import org.wordpress.android.fluxc.model.WCTopEarnerModel
 import org.wordpress.android.fluxc.store.WCStatsStore.StatsGranularity
+import org.wordpress.android.fluxc.store.WCStatsStore.StatsGranularity.DAYS
+import org.wordpress.android.fluxc.store.WCStatsStore.StatsGranularity.YEARS
+import java.time.LocalDateTime
 import javax.inject.Inject
 
 class DashboardFragment : TopLevelFragment(), DashboardContract.View, DashboardStatsListener,
@@ -171,7 +174,12 @@ class DashboardFragment : TopLevelFragment(), DashboardContract.View, DashboardS
         // Only update the order stats view if the new stats match the currently selected timeframe
         if (dashboard_stats.activeGranularity == granularity) {
             dashboard_stats.showErrorView(false)
-            dashboard_stats.updateView(revenueStats, salesStats, presenter.getStatsCurrency())
+
+            if (granularity == YEARS) {
+                dashboard_stats.updateView(fakeYearlyRevenueStats, fakeYearlySalesStats, presenter.getStatsCurrency())
+            } else {
+                dashboard_stats.updateView(revenueStats, salesStats, presenter.getStatsCurrency())
+            }
         }
     }
 
@@ -199,6 +207,16 @@ class DashboardFragment : TopLevelFragment(), DashboardContract.View, DashboardS
     }
 
     override fun showVisitorStats(visitorStats: Map<String, Int>, granularity: StatsGranularity) {
+        if (dashboard_stats.activeGranularity == DAYS) {
+            dashboard_stats.showVisitorStats(mapOf("2020-04-27" to 1387))
+            return
+        } else if (dashboard_stats.activeGranularity == YEARS) {
+            dashboard_stats.showVisitorStats(
+                mapOf("2020-01-01" to fakeYearlyVisitorStats.values.reduce() { a, b -> a + b })
+            )
+            return
+        }
+
         if (dashboard_stats.activeGranularity == granularity) {
             dashboard_stats.showVisitorStats(visitorStats)
         }
@@ -349,4 +367,36 @@ class DashboardFragment : TopLevelFragment(), DashboardContract.View, DashboardS
         AppPrefs.setShouldDisplayV4StatsAvailabilityBanner(false)
         showV4StatsAvailabilityBanner(false)
     }
+
+    // See
+    // https://github.com/woocommerce/woocommerce-ios/blob/ea22f8241d18fe4b4a21d0084dc906500452b7a2/WooCommerce/WooCommerceUITests/Mocks/mappings/stats/stats_orders-year.json#L27-L33
+    private val fakeYearlyRevenueStats: Map<String, Double> = {
+        val currentYear = LocalDateTime.now().year
+        mapOf(
+            (currentYear - 4).toString() to 4000.0,
+            (currentYear - 3).toString() to 8400.0,
+            (currentYear - 2).toString() to 12750.0,
+            (currentYear - 1).toString() to 17800.0,
+            currentYear.toString() to 20240.0
+        )
+    }()
+
+    // See
+    // https://github.com/woocommerce/woocommerce-ios/blob/ea22f8241d18fe4b4a21d0084dc906500452b7a2/WooCommerce/WooCommerceUITests/Mocks/mappings/stats/stats_orders-year.json#L36
+    // No runtime evaluation of the current year, because what's the chance that this fixture won't need changing in the
+    // next 5 years?!
+    private val fakeYearlySalesStats: Map<String, Int> = mapOf("2020" to 520)
+
+    // See
+    // https://github.com/woocommerce/woocommerce-ios/blob/ea22f8241d18fe4b4a21d0084dc906500452b7a2/WooCommerce/WooCommerceUITests/Mocks/mappings/stats/stats_visits-year.json#L29-L35
+    private val fakeYearlyVisitorStats: Map<String, Int> = {
+        val currentYear = LocalDateTime.now().year
+        mapOf(
+            (currentYear - 4).toString() to 500,
+            (currentYear - 3).toString() to 900,
+            (currentYear - 2).toString() to 1200,
+            (currentYear - 1).toString() to 1350,
+            currentYear.toString() to 1299
+        )
+    }()
 }
```

</details>

With that addition, these are the screenshots I get

<details>
<summary>Show image (it's <i>long</i>)</summary>

![screencapture-file-Users-gio-Developer-a8c-wcandroid-fastlane-screenshots-screenshots-html-2020-06-23-21_32_15](https://user-images.githubusercontent.com/1218433/85401122-d584d780-b59c-11ea-80a2-60aa8a6ebc14.png)


</details>

jkmassel commented:
>Although, the tr locale fails when running from Fastlane (but not from Android studio with -testLocale 'tr'). It looks like a timeout fetching the graph data.

I think this is probably a blocker, because if we wanted to do a full screenshot generation run, it'd fail. A couple of suggestions to get this PR merged:

1. We could just disable `tr` for now in the screenshot capture script. That language will just be stuck as-is until we get a fix done.
2. You could try adding an artificial delay while loading that specific page – see if it fixes the issue or not? If that doesn't work, you could potentially detect the failure and retry as well. 

For mocking the data – the tests already have their own mocked dagger graph (which we're not currently using). If we _were_ using it, we could inject our own class that can provide stats data – I suspect someone on the WooAndroid team could help that! :)

mokagio commented:
Closing this because it's embarrassingly stale.

I spoke with @jkmassel and we agree that enough has changed since working on this not to grant an update.

I've logged this in my personal record keeping. Hopefully the next time we'll work on screenshots automation for Android, I'll be able to share some of the learnings made here.
